src/dgipy/graph_app.py

```diff
@@ -1,218 +1,257 @@
-    @app.callback(
-        Output("edge-info-text", "children"),
-        [Input("selected-node", "data"), Input("neighbor-dropdown", "value")],
-        State("graph", "data"),
-    )
-    def update(
-        selected_node: str | dict,
-        selected_neighbor: str | None,
-        graph: dict | None,
-    ) -> str:
-        if selected_node == "":
-            return "No Edge Selected"
-        if selected_node["curveNumber"] == 1:
-            selected_data = __get_node_data_from_id(
-                graph["links"], selected_node["text"]
-            )
-            return (
-                "ID: "
-                + str(selected_data["id"])
-                + "\n\nApproval: "
-                + str(selected_data["approval"])
-                + "\n\nScore: "
-                + str(selected_data["score"])
-                + "\n\nAttributes: "
-                + str(selected_data["attributes"])
-                + "\n\nSource: "
-                + str(selected_data["source"])
-                + "\n\nPmid: "
-                + str(selected_data["pmid"])
-            )
-        if selected_neighbor is not None:
-            edge_node_id = None
-            selected_node_is_gene = __get_node_data_from_id(
-                graph["nodes"], selected_node["text"]
-            )["isGene"]
-            selected_neighbor_is_gene = __get_node_data_from_id(
-                graph["nodes"], selected_neighbor
-            )["isGene"]
-            if selected_node_is_gene == selected_neighbor_is_gene:
-                return dash.no_update
-            if selected_node_is_gene:
-                edge_node_id = selected_node["text"] + " - " + selected_neighbor
-            elif selected_neighbor_is_gene:
-                edge_node_id = selected_neighbor + " - " + selected_node["text"]
-            selected_data = __get_node_data_from_id(graph["links"], edge_node_id)
-            if selected_data is None:
-                return dash.no_update
-            return (
-                "ID: "
-                + str(selected_data["id"])
-                + "\n\nApproval: "
-                + str(selected_data["approval"])
-                + "\n\nScore: "
-                + str(selected_data["score"])
-                + "\n\nAttributes: "
-                + str(selected_data["attributes"])
-                + "\n\nSource: "
-                + str(selected_data["source"])
-                + "\n\nPmid: "
-                + str(selected_data["pmid"])
-            )
-        return "No Edge Selected"
-
-
-def __get_node_data_from_id(nodes: list, node_id: str) -> dict | None:
-    for node in nodes:
-        if node["id"] == node_id:
-            return node
-    return None
+"""Provides functionality to create a Dash web application for interacting with drug-gene data from DGIdb"""
+
+import dash_bootstrap_components as dbc
+from dash import Input, Output, State, ctx, dash, dcc, html
+
+from dgipy import dgidb
+from dgipy import network_graph as ng
+
+

```

Comment:
I'm not sure if this is a significant problem or anything, but I noticed that the first time you switch from gene mode to drug mode there is a long load time. Presumably this is just querying the drug list or something?

Reply:
Oddly enough, the long load time seems to be from dash updating the terms dropdown list with the drug list (since it is really long). The drug (and by extension gene) list query only is called once so it isn't that. I do want to come back around to this and see if there are any ways to fix the delay. Off the top of my head I have an idea wherein there are two dropdown menus, one for genes and drugs, so dash has no delay trying to update the list, and I just show each dropdown menu whenever appropriate. I do, however, want to see if there is a more elegant solution.